### ISSUE TITLE
Encrypt credentials in database

### ISSUE THREAD

**Issue by Phyks** [cantino/huginn#1843]
Hi,

Some agents require credentials, which could be an API key or even some real credentials. I am not 100% sure, but it seems they are stored in plaintext in a `options` text field in the `agents` table.

Storing these credentials in plaintext in the database could result in exposing them. I think these agents options are user-specific, then, what about encrypting them with the user password (or even better anything derived from the user password?).

It would just require to store in memory a one-way derivation of the user password when logging in, and decrypt the credentials / API keys using this variable.

What do you think about it?

Thanks!

P.S.: Some extra context: I am willing to get some "scraping" agents in Huginn to interact in an IFTTT way with these websites, typically backed by [Weboob/](http://weboob.org/), and this would result in options having my actual credentials for these specific websites, which does not seem to be a really good idea (typically for bank accounts).

**Comment by dsander** [comment]
Hi again,

encrypting the credentials (not the Agent options) sounds like a good idea. Did you come across the "Credentials" page of Huginn? It's basically a key-value store that can be accessed from any Agents options with a [liquid](https://github.com/cantino/huginn/wiki/Formatting-Events-using-Liquid) tag: `{% credential CREDENTIAL_NAME %}`.

For the encryption key we could use `APP_SECRET_TOKEN`, the users password does not work because the decryption also needs to work after restarting the application and we do not (and shouldn't) store the plain text password. 
The encryption of credentials would protect the database server, but if the application server is compromised it does not help.

**Comment by cantino** [comment]
Thinking through this: if we use`APP_SECRET_TOKEN`, then if the DB is hacked but the app isn't, the credentials are secure. If the app is hacked, there's no benefit to encrypting as the key must be stored in the app somehow and could thus be stolen. It seems more likely to me that the app would get hacked than the DB, so I'm not sure how much this helps real security, but I'm open to it if @Phyks wants to work on it. You could also consider using an HSM if you're deploying on AWS, but that'd be some custom integrations that I suspect few people would use.

**Comment by Phyks** [comment]
Indeed, I completely missed the credentials page (there does not seem to be any reference to it in the wiki :/). Looks like a perfect starting point!

From my perspective, I was considering a few scenarios:
* First scenario, if the database leaks for any reason (typically another app with too wide accesses and a vulnerability exposing the database). In such a case, I'd like the credentials to be stored in an encrypted way, and `APP_SECRET_TOKEN` should work really well.
* Second scenario is the `huginn` app itself being compromised.
* Third scenario is the whole server being compromised.

For the first one, `APP_SECRET_TOKEN` is enough. For the case where the app itself (or the server) is being compromised, I am not sure anything at all can be done. I know [CozyCloud](http://cozy.io/) encrypts everything in db with the user password, and unlock it at user login. The session is then valid for a week or so, which means credentials are unencrypted in memory for this period. However, this comes with its own problems, and I am not convinced it would work in the case of `huginn`.

Implementing anything AWS-specific seems really specific (plus I am self-hosting, without a HSM for my particular use case).

I can have a look at encrypting user credentials using `APP_SECRET_TOKEN` and decrypting on the fly, if you think it can be worth it and mitigate some potential security issues. Should it be done anyways, or triggered via a dotenv option?

**Comment by dsander** [comment]
One additional benefit would be that one could store database backups un-encrypted because the sensible information is encrypted inside.

Maybe it's better to use a new .env variable for the encryption key (which would also enable or disable the encryption). `APP_SECRET_TOKEN` could then still be changed to invalidate all user sessions.

**Comment by Phyks** [comment]
@cantino Sure, it might be worth encrypting more than the credentials.

I was thinking about using something like http://ruby-doc.org/stdlib-1.9.3/libdoc/openssl/rdoc/OpenSSL/Cipher.html with `AES-256-CBC`, unless anyone has a better idea.

There is also the question of the migration, which is to handle.

**Comment by Phyks** [comment]
I just came across https://github.com/attr-encrypted/attr_encrypted which seems to be a good and maintained solution to me. I'll try to submit a PoC in a PR soon.
